flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/offset/BinlogOffset.java

```diff
@@ -0,0 +1,266 @@
+
+    private final Map<String, String> offset;
+
+    public BinlogOffset(Map<String, String> offset) {
+        this.offset = offset;
+    }
+
+    public BinlogOffset(String filename, long position) {
+        this(filename, position, 0L, 0L, 0L, null, null);
+    }
+
+    public BinlogOffset(
+            String filename,
+            long position,
+            long restartSkipEvents,
+            long restartSkipRows,
+            long binlogEpochSecs,
+            @Nullable String restartGtidSet,
+            @Nullable Integer serverId) {
+        Map<String, String> offsetMap = new HashMap<>();
+        offsetMap.put(BINLOG_FILENAME_OFFSET_KEY, filename);
+        offsetMap.put(BINLOG_POSITION_OFFSET_KEY, String.valueOf(position));
+        offsetMap.put(EVENTS_TO_SKIP_OFFSET_KEY, String.valueOf(restartSkipEvents));
+        offsetMap.put(ROWS_TO_SKIP_OFFSET_KEY, String.valueOf(restartSkipRows));
+        offsetMap.put(TIMESTAMP_KEY, String.valueOf(binlogEpochSecs));
+        if (restartGtidSet != null) {
+            offsetMap.put(GTID_SET_KEY, restartGtidSet);
+        }
+        if (serverId != null) {
+            offsetMap.put(SERVER_ID_KEY, String.valueOf(serverId));
+        }
+        this.offset = offsetMap;
+    }
+
+    public Map<String, String> getOffset() {
+        return offset;
+    }
+
+    public String getFilename() {
+        return offset.get(BINLOG_FILENAME_OFFSET_KEY);
+    }
+
+    public long getPosition() {
+        return longOffsetValue(offset, BINLOG_POSITION_OFFSET_KEY);
+    }
+
+    public long getRestartSkipEvents() {
+        return longOffsetValue(offset, EVENTS_TO_SKIP_OFFSET_KEY);
+    }
+
+    public long getRestartSkipRows() {
+        return longOffsetValue(offset, ROWS_TO_SKIP_OFFSET_KEY);
+    }
+
+    public String getGtidSet() {
+        return offset.get(GTID_SET_KEY);
+    }
+
+    public long getTimestamp() {
+        return longOffsetValue(offset, TIMESTAMP_KEY);
+    }
+
+    public Long getServerId() {
+        return longOffsetValue(offset, SERVER_ID_KEY);
+    }
+
+    private long longOffsetValue(Map<String, ?> values, String key) {
+        Object obj = values.get(key);
+        if (obj == null) {
+            return 0L;
+        }
+        if (obj instanceof Number) {
+            return ((Number) obj).longValue();
+        }
+        try {
+            return Long.parseLong(obj.toString());
+        } catch (NumberFormatException e) {
+            throw new ConnectException(
+                    "Source offset '"
+                            + key
+                            + "' parameter value "
+                            + obj
+                            + " could not be converted to a long");
+        }
+    }
+
+    /**
+     * This method is inspired by {@link io.debezium.relational.history.HistoryRecordComparator}.
+     */
+    @Override
+    public int compareTo(BinlogOffset that) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 147-155, lines 157-194, lines 197-208 into separate methods.